### PR TITLE
[OptApp] Restructure 9 - Add vertex morphing

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This PR adds generic matrix based vertex morphing which can be used on variables following containers

1. Nodal historical variables
2. Nodal non historical variables
3. Elemental variables
4. Elemental properties variables
5. Condition variables
6. Condition properties variables

**🆕 Changelog**
- Added vertex morphing mappers
- Added tests

### CODE REVIEW

File: applications/OptimizationApplication/custom_utilities/mappers/container_variable_data_mapper.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        KRATOS_ERROR << "Calling ContainerVariableDataMapper::Update. This "
+                        "needs to be implemented in the derrived class.";
+    }
+
+    virtual void Map(
+        const ContainerVariableDataHolderType& rOriginDataContainer,
+        ContainerVariableDataHolderType& rDestinationDataContainer) const
+    {
+        KRATOS_ERROR << "Calling ContainerVariableDataMapper::Map. This needs to be "
+                        "implemented in the derrived class.";
```
Comment (FMeister) [inline applications/OptimizationApplication/custom_utilities/mappers/container_variable_data_mapper.h:62]:
```suggestion
                        "implemented in the derived class.";
```
Typo

File: applications/OptimizationApplication/custom_utilities/mappers/container_variable_data_mapper.h [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    virtual ~ContainerVariableDataMapper() = default;
+
+    ///@}
+    ///@name Public operations
+
+    virtual void Update()
+    {
+        KRATOS_ERROR << "Calling ContainerVariableDataMapper::Update. This "
+                        "needs to be implemented in the derrived class.";
```
Comment (FMeister) [inline applications/OptimizationApplication/custom_utilities/mappers/container_variable_data_mapper.h:54]:
```suggestion
                        "needs to be implemented in the derived class.";
```
Typo

File: applications/OptimizationApplication/custom_utilities/mappers/container_variable_data_mapper.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        KRATOS_ERROR << "Calling ContainerVariableDataMapper::Map. This needs to be "
+                        "implemented in the derrived class.";
+    }
+
+    virtual void InverseMap(
+        ContainerVariableDataHolderType& rOriginDataContainer,
+        const ContainerVariableDataHolderType& rDestinationDataContainer) const
+    {
+        KRATOS_ERROR << "Calling ContainerVariableDataMapper::Map. This needs to be "
+                        "implemented in the derrived class.";
```
Comment (FMeister) [inline applications/OptimizationApplication/custom_utilities/mappers/container_variable_data_mapper.h]:
```suggestion
                        "implemented in the derived class.";
```
Typo

File: applications/OptimizationApplication/custom_utilities/mappers/container_variable_data_mapper.h [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    virtual std::string Info() const
+    {
+        KRATOS_ERROR << "Calling ContainerVariableDataMapper::Info. This needs to be "
+                        "implemented in the derrived class.";
```
Comment (FMeister) [inline applications/OptimizationApplication/custom_utilities/mappers/container_variable_data_mapper.h]:
```suggestion
                        "implemented in the derived class.";
```
Typo

File: applications/OptimizationApplication/custom_utilities/mappers/vertex_morphing_container_variable_data_mapper.cpp [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+template<class TContainerType>
+void VertexMorphingContainerVariableDataMapper<TContainerType>::CreateSearchTreeWithAllNodesInOriginModelPart()
+{
+    mpSearchTree = Kratos::shared_ptr<KDTree>(new KDTree(mOriginEntityPointsVector.begin(), mOriginEntityPointsVector.end(), mBucketSize));
+}
+
+template<class TContainerType>
+void VertexMorphingContainerVariableDataMapper<TContainerType>::ComputeWeightForAllNeighbors(
+    double& rWieghtsSum,
```
Comment (FMeister) [inline applications/OptimizationApplication/custom_utilities/mappers/vertex_morphing_container_variable_data_mapper.cpp:107]:
```suggestion
    double& rWeightsSum,
```
Typo

File: applications/OptimizationApplication/python_scripts/transformation_techniques/transformation_technique.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import KratosMultiphysics as Kratos
+from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import ContainerVariableDataHolderUnion
+
+class TransformationTechnique(Kratos.Process):
+    def __init__(self):
+        super().__init__()
+
+    def TransformSensitivity(self, container_variable_data_holder: ContainerVariableDataHolderUnion):
+        raise NotImplementedError("Calling base class TransformationTechnique::TransformSensitivity. Please implement it in the derrived class")
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/transformation_techniques/transformation_technique.py]:
```suggestion
        raise NotImplementedError("Calling base class TransformationTechnique::TransformSensitivity. Please implement it in the derived class")
```
Typo

File: applications/OptimizationApplication/python_scripts/transformation_techniques/transformation_technique.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+    def TransformUpdate(self, container_variable_data_holder: ContainerVariableDataHolderUnion):
+        raise NotImplementedError("Calling base class TransformationTechnique::TransformUpdate. Please implement it in the derrived class")
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/transformation_techniques/transformation_technique.py]:
```suggestion
        raise NotImplementedError("Calling base class TransformationTechnique::TransformUpdate. Please implement it in the derived class")
```
Typo

File: applications/OptimizationApplication/tests/test_vertex_morphing_mappers.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
+from typing import Union
+from abc import ABC
+from abc import abstractmethod
+
+import KratosMultiphysics as Kratos
+import KratosMultiphysics.KratosUnittest as kratos_unittest
+import KratosMultiphysics.OptimizationApplication as KratosOA
+from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import ContainerVariableDataHolderUnion
+
+class TestVertexMoprhingContainerVariableDataMapper(ABC):
```
Comment (FMeister) [inline applications/OptimizationApplication/tests/test_vertex_morphing_mappers.py]:
```suggestion
class TestVertexMorphingContainerVariableDataMapper(ABC):
```
Typo

**Review by FMeister — COMMENTED** [review]
Besides my comments, it's fine

File: applications/OptimizationApplication/custom_utilities/mappers/vertex_morphing_container_variable_data_mapper.h [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+#include "custom_utilities/mappers/entity_point.h"
+#include "custom_utilities/mappers/mapping_filter_functions.h"
+
+namespace Kratos {
+
+///@name Kratos Classes
+///@{
+
+template<class TContainerType>
+class KRATOS_API(OPTIMIZATION_APPLICATION) VertexMorphingContainerVariableDataMapper : public ContainerVariableDataMapper<TContainerType>
```
Comment (sunethwarna) [inline applications/OptimizationApplication/custom_utilities/mappers/vertex_morphing_container_variable_data_mapper.h:38]: Rename the `mappers` folder to `filters`

File: applications/OptimizationApplication/python_scripts/transformation_techniques/transformation_technique.py (hunk 6 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/transformation_techniques/transformation_technique.py]: This will be moved to `filter`

**Comment by sunethwarna** [comment]
Closing since a better implementation is proposed in #11297

**Pull request closed without merging**